File: apps/api/src/routers/transactions.ts
import { TRPCError } from '@trpc/server';
import { createHash } from 'node:crypto';
import { and, between, desc, eq, gte, inArray, like, lte, sql } from 'drizzle-orm';
import { z } from 'zod';
import { db } from '@finances/db';
import { transactions, accounts, categories } from '@finances/db';
import {
  TransactionSchema,
  CreateTransactionInput,
  UpdateTransactionInput,
  ListTransactionsInput,
  BulkCreateInput,
  BulkCreateResult,
  RecentTransactionsInput,
  RecentTransactionSchema,
  SummaryByMonthInput,
  MonthlySummarySchema,
  SummaryByCategoryInput,
  CategoryAggregateSchema,
} from '@finances/contracts';
import { router, protectedProcedure } from '../trpc/trpc.js';

const IdInput = z.object({ id: z.string().uuid() });

function importHash(input: {
  accountId: string;
  date: string;
  amount: number;
  description: string;
}): string {
  return createHash('sha256')
    .update(`${input.accountId}|${input.date}|${input.amount}|${input.description}`)
    .digest('hex');
}

function monthKey(date: Date): string {
  const y = date.getUTCFullYear();
  const m = String(date.getUTCMonth() + 1).padStart(2, '0');
  return `${y}-${m}`;
}

function startOfMonth(monthKeyStr: string): Date {
  const [y, m] = monthKeyStr.split('-').map(Number);
  return new Date(Date.UTC(y ?? 1970, (m ?? 1) - 1, 1));
}

export const transactionsRouter = router({
  list: protectedProcedure
    .input(ListTransactionsInput)
    .output(z.array(TransactionSchema))
    .query(({ input }) => {
      const conditions = [];
      if (input.accountId) conditions.push(eq(transactions.accountId, input.accountId));
      if (input.categoryId) conditions.push(eq(transactions.categoryId, input.categoryId));
      if (input.kind) conditions.push(eq(transactions.kind, input.kind));
      if (input.search) conditions.push(like(transactions.description, `%${input.search}%`));
      if (input.from && input.to) {
        conditions.push(between(transactions.date, input.from, input.to));
      } else if (input.from) {
        conditions.push(gte(transactions.date, input.from));
      } else if (input.to) {
        conditions.push(lte(transactions.date, input.to));
      }

      const where = conditions.length > 0 ? and(...conditions) : undefined;

      return db
        .select()
        .from(transactions)
        .where(where)
        .orderBy(desc(transactions.date), desc(transactions.createdAt))
        .limit(input.limit)
        .offset(input.offset)
        .all();
    }),

  byId: protectedProcedure
    .input(IdInput)
    .output(TransactionSchema)
    .query(({ input }) => {
      const row = db.select().from(transactions).where(eq(transactions.id, input.id)).get();
      if (!row) {
        throw new TRPCError({ code: 'NOT_FOUND', message: 'Moviment no trobat' });
      }
      return row;
    }),

  create: protectedProcedure
    .input(CreateTransactionInput)
    .output(TransactionSchema)
    .mutation(({ input }) => {
      const id = crypto.randomUUID();
      const now = new Date().toISOString();
      db.insert(transactions)
        .values({
          id,
          accountId: input.accountId,
          categoryId: input.categoryId ?? null,
          kind: input.kind,
          amount: input.amount,
          description: input.description,
          notes: input.notes,
          date: input.date,
          importHash: null,
          transferAccountId: input.transferAccountId ?? null,
          createdAt: now,
          updatedAt: now,
        })
        .run();
      const row = db.select().from(transactions).where(eq(transactions.id, id)).get();
      if (!row) throw new Error('Failed to create transaction');
      return row;
    }),

  update: protectedProcedure
    .input(UpdateTransactionInput)
    .output(TransactionSchema)
    .mutation(({ input }) => {
      const { id, ...patch } = input;
      const existing = db.select().from(transactions).where(eq(transactions.id, id)).get();
      if (!existing) {
        throw new TRPCError({ code: 'NOT_FOUND', message: 'Moviment no trobat' });
      }
      const now = new Date().toISOString();
      db.update(transactions)
        .set({ ...patch, updatedAt: now })
        .where(eq(transactions.id, id))
        .run();
      const row = db.select().from(transactions).where(eq(transactions.id, id)).get();
      if (!row) throw new Error('Failed to update transaction');
      return row;
    }),

  delete: protectedProcedure
    .input(IdInput)
    .output(z.object({ id: z.string().uuid() }))
    .mutation(({ input }) => {
      db.delete(transactions).where(eq(transactions.id, input.id)).run();
      return { id: input.id };
    }),

  /**
   * True when there are zero transactions. Helps the UI distinguish
   * "fresh install" from "empty month".
   */
  hasAny: protectedProcedure
    .output(z.boolean())
    .query(() => {
      const row = db
        .select({ c: sql<number>`count(*)` })
        .from(transactions)
        .get();
      return (row?.c ?? 0) > 0;
    }),

  bulkCreate: protectedProcedure
    .input(BulkCreateInput)
    .output(BulkCreateResult)
    .mutation(({ input }) => {
      let created = 0;
      let skipped = 0;
      let errors = 0;
      const now = new Date().toISOString();

      for (const row of input.rows) {
        const hash = importHash({
          accountId: row.accountId,
          date: row.date,
          amount: row.amount,
          description: row.description ?? '',
        });
        try {
          const existing = db
            .select({ id: transactions.id })
            .from(transactions)
            .where(eq(transactions.importHash, hash))
            .get();
          if (existing) {
            skipped += 1;
            continue;
          }
          db.insert(transactions)
            .values({
              id: crypto.randomUUID(),
              accountId: row.accountId,
              categoryId: row.categoryId ?? null,
              kind: row.kind,
              amount: row.amount,
              description: row.description ?? '',
              notes: row.notes ?? '',
              date: row.date,
              importHash: hash,
              transferAccountId: row.transferAccountId ?? null,
              createdAt: now,
              updatedAt: now,
            })
            .run();
          created += 1;
        } catch {
          errors += 1;
        }
      }

      return { created, skipped, errors };
    }),

  recent: protectedProcedure
    .input(RecentTransactionsInput)
    .output(z.array(RecentTransactionSchema))
    .query(({ input }) => {
      const where =
        input.accountId !== undefined
          ? eq(transactions.accountId, input.accountId)
          : undefined;
      const rows = db
        .select({
          id: transactions.id,
          accountId: transactions.accountId,
          accountName: accounts.name,
          categoryId: transactions.categoryId,
          categoryName: categories.name,
          categoryColor: categories.color,
          kind: transactions.kind,
          amount: transactions.amount,
          description: transactions.description,
          date: transactions.date,
        })
        .from(transactions)
        .leftJoin(accounts, eq(transactions.accountId, accounts.id))
        .leftJoin(categories, eq(transactions.categoryId, categories.id))
        .where(where)
        .orderBy(desc(transactions.date), desc(transactions.createdAt))
        .limit(input.limit)
        .all();

      return rows.map((r) => ({
        id: r.id,
        accountId: r.accountId,
        accountName: r.accountName ?? 'Compte esborrat',
        categoryId: r.categoryId,
        categoryName: r.categoryName,
        categoryColor: r.categoryColor,
        kind: r.kind,
        amount: r.amount,
        description: r.description,
        date: r.date,
      }));
    }),

  summaryByMonth: protectedProcedure
    .input(SummaryByMonthInput)
    .output(z.array(MonthlySummarySchema))
    .query(({ input }) => {
      const today = new Date();
      const start = new Date(Date.UTC(today.getUTCFullYear(), today.getUTCMonth() - (input.months - 1), 1));
      const startStr = start.toISOString().slice(0, 10);

      const accountFilter =
        input.accountId !== undefined
          ? eq(transactions.accountId, input.accountId)
          : undefined;

      const rows = db
        .select({
          month: sql<string>`substr(${transactions.date}, 1, 7)`,
          income: sql<number>`COALESCE(SUM(CASE WHEN ${transactions.kind} = 'income' THEN ${transactions.amount} ELSE 0 END), 0)`,
          expense: sql<number>`COALESCE(SUM(CASE WHEN ${transactions.kind} = 'expense' THEN ${transactions.amount} ELSE 0 END), 0)`,
        })
        .from(transactions)
        .where(
          accountFilter
            ? and(gte(transactions.date, startStr), accountFilter)
            : gte(transactions.date, startStr),
        )
        .groupBy(sql`substr(${transactions.date}, 1, 7)`)
        .all();

      const byMonth = new Map<string, { incomeCents: number; expenseCents: number }>();
      for (const r of rows) {
        byMonth.set(r.month, { incomeCents: r.income, expenseCents: r.expense });
      }

      const out: { month: string; incomeCents: number; expenseCents: number; netCents: number }[] = [];
      for (let i = input.months - 1; i >= 0; i -= 1) {
        const d = new Date(Date.UTC(today.getUTCFullYear(), today.getUTCMonth() - i, 1));
        const key = monthKey(d);
        const totals = byMonth.get(key) ?? { incomeCents: 0, expenseCents: 0 };
        // Amounts are signed cents (expense < 0, income > 0). Normalise to
        // absolute for display and compute net from the raw signs.
        out.push({
          month: key,
          incomeCents: Math.abs(totals.incomeCents),
          expenseCents: Math.abs(totals.expenseCents),
          netCents: totals.incomeCents + totals.expenseCents,
        });
      }
      return out;
    }),

  summaryByCategory: protectedProcedure
    .input(SummaryByCategoryInput)
    .output(z.array(CategoryAggregateSchema))
    .query(({ input }) => {
      const conditions = [
        gte(transactions.date, input.from),
        lte(transactions.date, input.to),
      ];
      if (input.accountId !== undefined) {
        conditions.push(eq(transactions.accountId, input.accountId));
      }

      const rows = db
        .select({
          categoryId: transactions.categoryId,
          name: categories.name,
          color: categories.color,
          cents: sql<number>`COALESCE(SUM(${transactions.amount}), 0)`,
          count: sql<number>`count(*)`,
        })
        .from(transactions)
        .leftJoin(categories, eq(transactions.categoryId, categories.id))
        .where(and(...conditions))
        .groupBy(transactions.categoryId, categories.name, categories.color)
        .all();

      return rows.map((r) => ({
        categoryId: r.categoryId,
        name: r.name ?? 'Sense categoria',
        color: r.color ?? '#8B7355',
        cents: Math.abs(r.cents),
        count: r.count,
      }));
    }),
});

File: backend/internal/db/transactions.go
package db

import (
	"crypto/sha256"
	"database/sql"
	"encoding/hex"
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/mgurt/finances/internal/models"
)

// TransactionsStore provides query methods for transactions.
type TransactionsStore struct {
	*Store
}

// TransactionFilters holds all optional filter parameters for List.
type TransactionFilters struct {
	AccountID   string
	CategoryID  string
	Kind        string
	Search      string
	From        string
	To          string
	Limit       int
	Offset      int
}

// List returns transactions matching the given filters, ordered by date desc, createdAt desc.
func (t *TransactionsStore) List(filters TransactionFilters) ([]models.Transaction, error) {
	where, args := buildTransactionWhereClause(filters)

	query := "SELECT id, account_id, category_id, kind, amount, description, notes, date, import_hash, transfer_account_id, created_at, updated_at FROM transactions"
	if where != "" {
		query += " WHERE " + where
	}
	query += " ORDER BY date DESC, created_at DESC"
	if filters.Limit > 0 {
		query += fmt.Sprintf(" LIMIT %d", filters.Limit)
	}
	if filters.Offset > 0 {
		query += fmt.Sprintf(" OFFSET %d", filters.Offset)
	}

	rows, err := t.DB.Query(query, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var transactions []models.Transaction
	for rows.Next() {
		var tx models.Transaction
		err := rows.Scan(
			&tx.ID, &tx.AccountID, &tx.CategoryID, &tx.Kind, &tx.Amount,
			&tx.Description, &tx.Notes, &tx.Date, &tx.ImportHash,
			&tx.TransferAccountID, &tx.CreatedAt, &tx.UpdatedAt,
		)
		if err != nil {
			return nil, err
		}
		transactions = append(transactions, tx)
	}
	return transactions, rows.Err()
}

func buildTransactionWhereClause(filters TransactionFilters) (string, []interface{}) {
	var conditions []string
	var args []interface{}

	if filters.AccountID != "" {
		conditions = append(conditions, "account_id = ?")
		args = append(args, filters.AccountID)
	}
	if filters.CategoryID != "" {
		conditions = append(conditions, "category_id = ?")
		args = append(args, filters.CategoryID)
	}
	if filters.Kind != "" {
		conditions = append(conditions, "kind = ?")
		args = append(args, filters.Kind)
	}
	if filters.Search != "" {
		conditions = append(conditions, "(description LIKE ? OR notes LIKE ?)")
		searchPattern := "%" + filters.Search + "%"
		args = append(args, searchPattern, searchPattern)
	}
	if filters.From != "" {
		conditions = append(conditions, "date >= ?")
		args = append(args, filters.From)
	}
	if filters.To != "" {
		conditions = append(conditions, "date <= ?")
		args = append(args, filters.To)
	}

	if len(conditions) == 0 {
		return "", nil
	}
	return strings.Join(conditions, " AND "), args
}

// ByID returns a single transaction by ID.
func (t *TransactionsStore) ByID(id string) (*models.Transaction, error) {
	query := "SELECT id, account_id, category_id, kind, amount, description, notes, date, import_hash, transfer_account_id, created_at, updated_at FROM transactions WHERE id = ?"
	var tx models.Transaction
	err := t.DB.QueryRow(query, id).Scan(
		&tx.ID, &tx.AccountID, &tx.CategoryID, &tx.Kind, &tx.Amount,
		&tx.Description, &tx.Notes, &tx.Date, &tx.ImportHash,
		&tx.TransferAccountID, &tx.CreatedAt, &tx.UpdatedAt,
	)
	if err == sql.ErrNoRows {
		return nil, fmt.Errorf("transaction not found")
	}
	if err != nil {
		return nil, err
	}
	return &tx, nil
}

// Create inserts a new transaction and returns it.
func (t *TransactionsStore) Create(req models.CreateTransactionReq) (*models.Transaction, error) {
	id := uuid.New().String()
	now := time.Now().UTC().Format(time.RFC3339)
	description := req.Description
	if description == "" {
		description = ""
	}
	notes := req.Notes
	if notes == "" {
		notes = ""
	}

	_, err := t.DB.Exec(`
		INSERT INTO transactions (id, account_id, category_id, kind, amount, description, notes, date, import_hash, transfer_account_id, created_at, updated_at)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		id, req.AccountID, req.CategoryID, req.Kind, req.Amount, description, notes,
		req.Date, req.ImportHash, req.TransferAccountID, now, now)
	if err != nil {
		return nil, err
	}

	return t.ByID(id)
}

// Update modifies an existing transaction and returns it.
func (t *TransactionsStore) Update(id string, req models.UpdateTransactionReq) (*models.Transaction, error) {
	existing, err := t.ByID(id)
	if err != nil {
		return nil, err
	}

	categoryID := req.CategoryID
	if categoryID == nil {
		categoryID = existing.CategoryID
	}
	kind := req.Kind
	if kind == "" {
		kind = existing.Kind
	}
	amount := req.Amount
	if amount == 0 {
		amount = existing.Amount
	}
	description := req.Description
	if description == "" {
		description = existing.Description
	}
	notes := req.Notes
	if notes == "" {
		notes = existing.Notes
	}
	date := req.Date
	if date == "" {
		date = existing.Date
	}
	transferAccountID := req.TransferAccountID
	if transferAccountID == nil {
		transferAccountID = existing.TransferAccountID
	}
	now := time.Now().UTC().Format(time.RFC3339)

	_, err = t.DB.Exec(`
		UPDATE transactions SET category_id=?, kind=?, amount=?, description=?, notes=?, date=?, transfer_account_id=?, updated_at=? WHERE id=?`,
		categoryID, kind, amount, description, notes, date, transferAccountID, now, id)
	if err != nil {
		return nil, err
	}

	return t.ByID(id)
}

// Delete removes a transaction by ID.
func (t *TransactionsStore) Delete(id string) error {
	_, err := t.ByID(id)
	if err != nil {
		return err
	}
	_, err = t.DB.Exec("DELETE FROM transactions WHERE id = ?", id)
	return err
}

// BulkDelete removes multiple transactions in a single SQL statement.
// Returns the number of rows actually deleted (may be < len(ids) if some
// IDs no longer exist). The single-statement DELETE is atomic at the
// SQLite level — either all rows matching the IN clause are removed, or
// none (in the event of a constraint violation).
func (t *TransactionsStore) BulkDelete(ids []string) (int, error) {
	if len(ids) == 0 {
		return 0, nil
	}
	placeholders := make([]string, len(ids))
	args := make([]any, len(ids))
	for i, id := range ids {
		placeholders[i] = "?"
		args[i] = id
	}
	query := "DELETE FROM transactions WHERE id IN (" + strings.Join(placeholders, ",") + ")"
	res, err := t.DB.Exec(query, args...)
	if err != nil {
		return 0, err
	}
	n, err := res.RowsAffected()
	if err != nil {
		return 0, err
	}
	return int(n), nil
}

// HasAny returns true if any transactions exist.
func (t *TransactionsStore) HasAny() (bool, error) {
	var count int
	err := t.DB.QueryRow("SELECT COUNT(*) FROM transactions").Scan(&count)
	if err != nil {
		return false, err
	}
	return count > 0, nil
}

// BulkCreate inserts multiple transactions, skipping those with importHash duplicates.
// Returns the count of actually inserted transactions.
func (t *TransactionsStore) BulkCreate(req models.BulkCreateReq) (int, error) {
	if len(req.Transactions) == 0 {
		return 0, nil
	}

	// Collect all importHashes to check for existing duplicates
	existingHashes := make(map[string]bool)
	for _, tx := range req.Transactions {
		if tx.ImportHash != nil && *tx.ImportHash != "" {
			// SHA256 the import hash as the DB stores it
			hash := sha256.Sum256([]byte(*tx.ImportHash))
			hashStr := hex.EncodeToString(hash[:])
			var count int
			err := t.DB.QueryRow("SELECT COUNT(*) FROM transactions WHERE import_hash = ?", hashStr).Scan(&count)
			if err != nil {
				return 0, err
			}
			if count > 0 {
				existingHashes[*tx.ImportHash] = true
			}
		}
	}

	now := time.Now().UTC().Format(time.RFC3339)
	inserted := 0

	for _, item := range req.Transactions {
		if item.ImportHash != nil && existingHashes[*item.ImportHash] {
			continue // skip duplicate
		}

		id := uuid.New().String()
		var importHashStr *string
		if item.ImportHash != nil && *item.ImportHash != "" {
			hash := sha256.Sum256([]byte(*item.ImportHash))
			s := hex.EncodeToString(hash[:])
			importHashStr = &s
		}

		_, err := t.DB.Exec(`
			INSERT INTO transactions (id, account_id, category_id, kind, amount, description, notes, date, import_hash, transfer_account_id, created_at, updated_at)
			VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
			id, item.AccountID, item.CategoryID, item.Kind, item.Amount,
			item.Description, item.Notes, item.Date, importHashStr, item.TransferAccountID, now, now)
		if err != nil {
			return inserted, err
		}
		inserted++
	}

	return inserted, nil
}

// Recent returns the most recent transactions with account and category names.
func (t *TransactionsStore) Recent(limit int) ([]models.TransactionWithDetails, error) {
	if limit == 0 {
		limit = 10
	}

	query := `
		SELECT
			t.id, t.account_id, t.category_id, t.kind, t.amount, t.description, t.notes,
			t.date, t.import_hash, t.transfer_account_id, t.created_at, t.updated_at,
			a.name AS account_name,
			COALESCE(c.name, '') AS category_name,
			COALESCE(ta.name, '') AS transfer_account_name
		FROM transactions t
		JOIN accounts a ON a.id = t.account_id
		LEFT JOIN categories c ON c.id = t.category_id
		LEFT JOIN accounts ta ON ta.id = t.transfer_account_id
		ORDER BY t.date DESC, t.created_at DESC
		LIMIT ?`

	rows, err := t.DB.Query(query, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var transactions []models.TransactionWithDetails
	for rows.Next() {
		var tx models.TransactionWithDetails
		err := rows.Scan(
			&tx.ID, &tx.AccountID, &tx.CategoryID, &tx.Kind, &tx.Amount,
			&tx.Description, &tx.Notes, &tx.Date, &tx.ImportHash,
			&tx.TransferAccountID, &tx.CreatedAt, &tx.UpdatedAt,
			&tx.AccountName, &tx.CategoryName, &tx.TransferAccountName,
		)
		if err != nil {
			return nil, err
		}
		transactions = append(transactions, tx)
	}
	return transactions, rows.Err()
}

// SummaryByMonth returns income, expense, net, and count per month.
func (t *TransactionsStore) SummaryByMonth(months int) ([]models.SummaryByMonthItem, error) {
	if months == 0 {
		months = 12
	}

	query := `
		SELECT
			strftime('%Y-%m', date) AS month,
			COALESCE(SUM(CASE WHEN amount > 0 THEN amount ELSE 0 END), 0) AS income,
			COALESCE(SUM(CASE WHEN amount < 0 THEN ABS(amount) ELSE 0 END), 0) AS expense,
			COALESCE(SUM(amount), 0) AS net,
			COUNT(*) AS count
		FROM transactions
		GROUP BY strftime('%Y-%m', date)
		ORDER BY month DESC
		LIMIT ?`

	rows, err := t.DB.Query(query, months)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var summaries []models.SummaryByMonthItem
	for rows.Next() {
		var s models.SummaryByMonthItem
		err := rows.Scan(&s.Month, &s.Income, &s.Expense, &s.Net, &s.Count)
		if err != nil {
			return nil, err
		}
		summaries = append(summaries, s)
	}
	return summaries, rows.Err()
}

// SummaryByCategory returns total and count per category for a date range.
func (t *TransactionsStore) SummaryByCategory(from, to string) ([]models.SummaryByCategoryItem, error) {
	query := `
		SELECT
			t.category_id,
			COALESCE(c.name, 'Uncategorized') AS category_name,
			COALESCE(SUM(ABS(t.amount)), 0) AS total,
			COUNT(*) AS count
		FROM transactions t
		LEFT JOIN categories c ON c.id = t.category_id
		WHERE t.kind = 'expense' AND t.date >= ? AND t.date <= ?
		GROUP BY t.category_id
		ORDER BY total DESC`

	rows, err := t.DB.Query(query, from, to)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var summaries []models.SummaryByCategoryItem
	for rows.Next() {
		var s models.SummaryByCategoryItem
		err := rows.Scan(&s.CategoryID, &s.CategoryName, &s.Total, &s.Count)
		if err != nil {
			return nil, err
		}
		summaries = append(summaries, s)
	}
	return summaries, rows.Err()
}